### ISSUE TITLE
service bus binding does not accept EntityPath as connect string

### ISSUE THREAD

**Issue by lindydonna** [Azure/azure-functions-servicebus-extension#25]
_From @globysdz on May 26, 2017 18:48_

{
  "disabled": false,
  "bindings": [
    {
      "name" : "myFuncname",
      "topicName" : "request",
      "subscriptionName" : "processor",
      "connection" : "IncomingSubscription"
      "accessRights" : "listen",
      "type" : "serviceBusTrigger",
      "direction" : "in"
    }
  ]
}

the "incomingSubscription" was set via ARM template: 
    {
              "name": "IncomingSubscription",
              "value": "[listkeys(resourceId('Microsoft.ServiceBus/namespaces/topics/authorizationRules', parameters('bus_name'), 'request', 'perm-r'), variables('sbVersion')).primaryConnectionString]" 
     }

the result "inccomingSubscription" string contains ";EntityPath=xxxx" at the end of the string which AzureFunction throws exception at startup time:

Error indexing method 'Functions.globyscadProcessor' ---> System.ArgumentException : ConnectionString used to create MessagingFactory shouldn't include EntityPath as MessagingFactory is at Namespace level.

I am using the same connection string for my webapi application, which works fine. but for function it doesn't work. and there is no good way to automatically set the connection string variable (unless there is a way to obtain the connection string from ARM without EntityPath=xxx part).

_Copied from original issue: Azure/Azure-Functions#315_

**Comment by dougmolineux** [comment]
To sum it up, in order to use topic-level connection strings, we have no alternative other than a string operation to "chop" off the `EntityPath=`? This sounds a bit hacky.

**Comment by ThorstenHans** [comment]
It's not just hacky @dougmolineux if you, for example, use https://keda.sh to run Azure Functions in Kubernetes, you cant use Azure ServiceBus as Scaler, because KEDA requires the `EntityPath` in the ConnectionString. 

This has to be unified somehow

**Comment by ericsampson** [comment]
@alrod your [commit](https://github.com/Azure/azure-functions-servicebus-extension/commit/791274ecc5d334dd3652fad83c71cbf19e916bdb) stated it fixed this issue, but it remains open. Is it actually fixed and this issue should be closed? Thanks.

**Comment by shivanigoyaluk** [comment]
I am using Microsoft.Azure.WebJobs.Extentions.ServiceBus 4.1.1 and still got the issue.

**Comment by asears** [comment]
Still seeing this too, the EntityPath should be optional and not complain in the SDK, as it is the default connection string coming out of the Azure Service Bus Service.  

There are some limitations with Service Bus input/output connection strings in Azure Functions and this is problematic for key rotation/duplication of keys with KEDA.

**Comment by bojanfilipovic** [comment]
I've also encountered this, tried using Microsoft.Azure.WebJobs.Extensions.ServiceBus versions 4.1.1 and 4.1.2.
Any news on when this will be fixed?

**Comment by asears** [comment]
> To sum it up, in order to use topic-level connection strings, we have no alternative other than a string operation to "chop" off the `EntityPath=`? This sounds a bit hacky.

This becomes more problematic when secret keys include the delimiter.

**Comment by antempus** [comment]
Any update on this?

**Comment by jefking** [comment]
This is not good. We pull our keys out of Service Bus via ARM and store them in Key Vault; we do not have the ability to 'hack' through our environments, due to compliance. What is a reasonable workaround, if this is an open issue that has been going on for years.

**Comment by mcalnd70** [comment]
@jefking I'm using Terraform (but I would imagine ARM has a similar function to this) and the code that stores the Key Vault secrets from Service Bus infrastructure provision puts in x2 different KV Secrets, one with the full connection string, the other with the Entity Path stripped off, just for Function Apps usage.

For example...This is Terraform code... the main object of interest here is the **trimsuffix** part...
(the **_$var.azsbtopic1_** on the end of the trim would need to be the actual name of your SB topic or element, named explicitly for this to work)

`resource "azurerm_key_vault_secret" "go-1" {
  name          = "keyvaultsecret-MyServiceBus-PrimaryConnectionString"
  value         = azurerm_servicebus_topic_authorization_rule.go-1.primary_connection_string
  key_vault_id  = data.azurerm_key_vault.env-keyvault.id
}

# No EntityPath versions of the Primary connection strings outside of the namespace - this is for Function Apps to connect successfully
# https://github.com/Azure/azure-functions-servicebus-extension/issues/25
resource "azurerm_key_vault_secret" "go-1a" {
  name          = "keyvaultsecret-MyServiceBus-PrimaryConnectionString-NoEntityPath"
  value         = trimsuffix("${azurerm_servicebus_topic_authorization_rule.go-1.primary_connection_string}", ";EntityPath=${var.azsbtopic1}")  
  key_vault_id  = data.azurerm_key_vault.env-keyvault.id
}`

The tl;dr version of this - Trim off the ";EntityPath=whatever" in the connection string and store it separately in your Key Vault as a different named secret, and use that one for your Function Apps.

**Comment by jefking** [comment]
I was thinking about this as a solution as well. Thanks for posting; it should be an idea for anyone in this position!
As I was in a rush, I ended up leveraging a 'send' auth token, which works from all the topics/queues...
I'll come back to this soon, and post the ARM solution, if we come up with one.

**Comment by bjorn-jarisch** [comment]
Instead of trying to trim the connection string, it might be easier to _create_ the connection string without the EntityPath using the primary key.

Here's how we get a function-app-compatible connection string in an ARM template:

```
{
  "name": "ServiceBusConnectionString",
  "value": "[concat('Endpoint=sb://', variables('serviceBusName'), '.servicebus.windows.net/;SharedAccessKeyName=', variables('keyName'), ';SharedAccessKey=', listKeys(resourceId(variables('serviceBusResourceGroup'), 'Microsoft.ServiceBus/namespaces/topics/authorizationRules', variables('serviceBusName'), variables('topicName'), variables('keyName')),'2015-08-01').primaryKey)]"
}
```

**Comment by Fabiest** [comment]
I just attempted to create a simple branch with some very simple changes on this but I don't have permissions. My bad.

Any way, since I don't understand much about bindings and somewhat the overall usage of a lot of the code, I went straight forward to what I need: Ability to provide, in an azure function using the ServiceBusTrigger, connection string with entity path.

My changes can be found attached to this and were only on class MessagingProvider.
[MessagingProvider.zip](https://github.com/Azure/azure-functions-servicebus-extension/files/5922774/MessagingProvider.zip)

Basic idea was to, everywhere where a new Microsoft.Azure.ServiceBus.ClientEntity is being created, instead of using the constructor which receives the connection string and queue/topic name, it now uses ServiceBusConnectionStringBuilder (based on the connection string). I also added some logic to check that the entity path, if present, in the connection string was the same as the one provided in the ServiceBusTrigger.

I'm certain that many things are not ok with my changes like: 

- The exception I throw when entity path provided vs conn. string are different.
- The fact that you still need to provide a entity path even if your conn. string already has it.
- Possibly missing some larger context explaining why such simple change is to valid.
- etc.

But my idea was to see if this gets, hopefully, picked up with some head start?

I mean... for my case I'm now forced to create a change in a different repo where my service bus exists (ARM) and it creates the Listen access conn. string into a Key Vault so it then works properly with this function. Worst case I will have to have 2 secrets in key vault for the same thing, one with entity path and the other without, because for me, the one without is the exception.

The only good point in what I did is that I tested, for my case, and it seemed to work fine 😋

**Comment by alrod** [comment]
This is fixed in [latest](https://www.nuget.org/packages/Microsoft.Azure.WebJobs.Extensions.ServiceBus/5.5.0) ServiceBus extension. Note that entity name must match in `ServiceBusTrigger` ctr and `EntityName` in in connection string otherwise you will get "The listener for function 'Function1' was unable to start. Azure.Messaging.ServiceBus: The queue or topic name provided does not match the EntityPath in the connection string passed to the ServiceBusClient constructor."

**Comment by filip-be** [comment]
Hi @alrod,
We are still facing the issue despite using the latest ServiceBus extension (5.6.0).

Logged error:
```
Unable to create NamespaceManager with ConnectionString in ServiceBusClientProvider.TryCreateNamespaceManager. Exception: System.ArgumentException: ConnectionString used to create NamespaceManager shouldn't include EntityPath as NamespaceManager is at Namespace level.
Parameter name: connectionString
at Microsoft.ServiceBus.NamespaceManager.CreateFromConnectionString(String connectionString)
at Microsoft.Azure.AppService.ScaleController.Extension.ClientProviders.External.ServiceBusClientProvider.TryCreateNamespaceManager(ServiceBusConnectionConfiguration connectionOptions, GetAadToken tokenFunc, NamespaceManager& namespaceManager, ILogger logger)
```

Topic name provided to the `ServiceBusTrigger` matches the `EntityPath` and the connection string and subscription name is retrieved from the App Function Config.

---

Connection string `Endpoint=sb://<sb name>.servicebus.windows.net/;SharedAccessKeyName=pipeline-state-changed-listen;SharedAccessKey=<key>=;EntityPath=pipeline-state-changed`
Function constructor: `public async Task RunAsync([ServiceBusTrigger("pipeline-state-changed", "%SubscriptionLogPipelineRun%", Connection = "pipelineStageChanged")] string pipelineData)`

**Comment by filip-be** [comment]
@alrod it seems to be even more confusing - when I have removed the `EntityPath` from the connection string then the error stopped from happening on a **serverless function**, but on a function with regular App Service plan I have started receiving below error:
```
The listener for function 'LogPipelineRun' was unable to start.
```

This is about the same function package, deployed in a different resource groups, with a different Function App.

**Comment by atif1996** [comment]
Is the workaround to remove the entity path?  Does removing the path change give the connection access other topics, or are the permissions still scoped to the topic the connection string was created for?
